File: src/main/java/jar/QuestionParser.java
package jar;

import edu.stanford.nlp.pipeline.*;

import java.io.*;
import java.nio.file.Files;
import java.nio.file.Path;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.util.*;
import edu.stanford.nlp.ling.*;
import edu.stanford.nlp.util.CoreMap;

import javax.xml.bind.DatatypeConverter;


public class QuestionParser {
    //The question parser will both be responsible for all of the NLP, and for caching the result
    //A question parser will be an object that is used only once

    ArrayList<String> parse = new ArrayList<String>();

    private Boolean WriteCache(ArrayList parsed) throws Exception {
        throw new Exception("Not Implemented");
    }

    private Boolean ReadCache(ArrayList parsed) throws Exception {
        throw new Exception("Not Implemented");
    }

    private void NLPParse(String question) {
        // creates a StanfordCoreNLP object, with POS tagging, lemmatization, NER, parsing, and coreference resolution
        Properties props = new Properties();
        props.setProperty("annotators", "tokenize, ssplit, pos, lemma, ner, parse");
        StanfordCoreNLP pipeline = new StanfordCoreNLP(props);

        // read some text in the text variable
        String text = question;

        // create an empty Annotation just with the given text
        Annotation document = new Annotation(text);

        // run all Annotators on this text
        pipeline.annotate(document);

        // these are all the sentences in this document
        // a CoreMap is essentially a Map that uses class objects as keys and has values with custom types
        List<CoreMap> sentences = document.get(CoreAnnotations.SentencesAnnotation.class);

        for(CoreMap sentence: sentences) {
            // traversing the words in the current sentence
            // a CoreLabel is a CoreMap with additional token-specific methods
            for (CoreLabel token: sentence.get(CoreAnnotations.TokensAnnotation.class)) {
                // this is the text of the token
                String word = token.get(CoreAnnotations.TextAnnotation.class);
                // this is the POS tag of the token
                String pos = token.get(CoreAnnotations.PartOfSpeechAnnotation.class);
                //extract the nouns

                if (pos.equals("NN") || pos.equals("NNP") || pos.equals("NNS") || pos.equals("NNPS")) {
                    parse.add(word);
                }
            }
        }

    }

    private String GenerateHash(String str) {
        MessageDigest md = null;
        try {
            md = MessageDigest.getInstance("MD5");
        } catch (NoSuchAlgorithmException e) {
            e.printStackTrace();
        }
        md.update(str.getBytes());
        byte[] digest = md.digest();
        String myHash = DatatypeConverter
                .printHexBinary(digest).toUpperCase();
        return myHash;
    }

    public QuestionParser (String question) {
        //Test if a file exists
        String pathString = "C:\\Users\\cdilg\\Documents\\dev\\SQLFitness\\cache\\" + GenerateHash(question) + ".q";
        //Path path = (pathString);
        File cacheQuestion = new File(pathString);
        if (cacheQuestion.exists()){
            try {
                Scanner cacheReader = new Scanner(cacheQuestion);
                while (cacheReader.hasNextLine()) {
                    parse.add(cacheReader.nextLine());
                }
            } catch (FileNotFoundException e) {
                e.printStackTrace();
            }

        } else {
            NLPParse(question);
            //Write to cache

            try {
                FileWriter cacheWriter = new FileWriter(cacheQuestion);
                for(int i =0; i < parse.size(); i++) {
                    cacheWriter.write(parse.get(i) + "\n");
                }
                cacheWriter.close();

            } catch (IOException e) {
                e.printStackTrace();
            }
        }
        //Read from cache
        //Else do the parse then



    }

    public ArrayList keywords() {
        return parse;
    }
}

File: SQLFitness/QuestionParser.cs
using edu.stanford.nlp.ling;
using edu.stanford.nlp.pipeline;
using edu.stanford.nlp.util;
using java.util;
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using static edu.stanford.nlp.ling.CoreAnnotations;

namespace SQLFitness
{
    class QuestionParser
    {
        public static List<string> ParseWithCache(string question) => parseQuestion(question);

        readonly static java.lang.Class sentencesAnnotationClass =
            new SentencesAnnotation().getClass();
        readonly static java.lang.Class tokensAnnotationClass =
            new TokensAnnotation().getClass();
        readonly static java.lang.Class textAnnotationClass =
            new TextAnnotation().getClass();
        readonly static java.lang.Class partOfSpeechAnnotationClass =
            new PartOfSpeechAnnotation().getClass();
        readonly static java.lang.Class namedEntityTagAnnotationClass =
            new NamedEntityTagAnnotation().getClass();
        readonly static java.lang.Class normalizedNamedEntityTagAnnotation =
            new NormalizedNamedEntityTagAnnotation().getClass();


        public static List<string> NLPParse(string question)
        {
            // creates a StanfordCoreNLP object, with POS tagging, lemmatization, NER, parsing, and coreference resolution
            Properties props = new Properties();
            props.setProperty("annotators", "tokenize, ssplit, pos, lemma, ner, parse");
            props.setProperty("ner.useSUTime", "0");
            var curDir = Environment.CurrentDirectory;
            var jarRoot = @"..\..\..\stanford-corenlp-full-2018-10-05\models";
            Directory.SetCurrentDirectory(jarRoot);
            StanfordCoreNLP pipeline = new StanfordCoreNLP(props);
            Directory.SetCurrentDirectory(curDir);

            // read some text in the text variable
            string text = question;

            // create an empty Annotation just with the given text
            Annotation document = new Annotation(text);

            // run all Annotators on this text
            pipeline.annotate(document);

            // these are all the sentences in this document
            var sentences = document.get(sentencesAnnotationClass) as AbstractList;

            var parse = new List<string>();

            foreach (CoreMap sentence in sentences)
            {
                // traversing the words in the current sentence
                // a CoreLabel has additional token-specific methods
                foreach (CoreLabel token in sentence.get(tokensAnnotationClass) as AbstractList)
                {
                    // this is the text of the token
                    string word = token.get(textAnnotationClass) as string;
                    //this is the POS tag of the token
                    string pos = token.get(partOfSpeechAnnotationClass) as string;
                    //extract the nouns

                    if (pos == "NN" || pos == "NNP" || pos == "NNS" || pos == "NNPS")
                    {
                        parse.Add(word);
                    }
                }
            }
            return parse;
        }

        private static List<string> parseQuestion(string question)
        {
            string pathString = @"..\..\..\cache\" + question.GetHashCode() + ".q";
            if (File.Exists(pathString))
            {
                return File.ReadAllLines(pathString).ToList();
            }
            else
            {
                var parsedQuestion = NLPParse(question);
                //Write to cache
                File.WriteAllLines(pathString, parsedQuestion);
                return parsedQuestion;
            }
        }
    }
}